File: 57. Insert Interval.cc
//TC: O(row)
//SC: O(1)
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        int row = intervals.size();
        vector<vector<int>> res;
        int curr = 0;
        for (int i = 0; i < row; i++) {
            if (intervals[i][1] < newInterval[0]) {
                res.push_back(intervals[i]);
                ++curr;
            } else if (intervals[i][0] > newInterval[1]) {
                res.push_back(intervals[i]);
            } else {
                newInterval[0] = min(newInterval[0], intervals[i][0]);
                newInterval[1] = max(newInterval[1], intervals[i][1]);
            }
        }
        res.insert(res.begin() + curr, newInterval);
        return res;
    }
};

File: java/57.insert_interval.java
// 57. Insert Interval
// TC: O(n)
// SC: O(n)
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int row = intervals.length;
        int i = 0;
        List<int[]> res = new ArrayList<>();
        int inserted_index = 0;
        for (; i < row; i++) {
            if (newInterval[0] > intervals[i][1]) {
                ++inserted_index;
                res.add(intervals[i]);
            } else if (newInterval[1] >= intervals[i][0]) {
                newInterval[0] = newInterval[0] < intervals[i][0] ? newInterval[0] : intervals[i][0];
                newInterval[1] = newInterval[1] > intervals[i][1] ? newInterval[1] : intervals[i][1];
            } else {
                res.add(intervals[i]);
            }
        }
        res.add(inserted_index, newInterval);
        return res.toArray(new int[res.size()][]);
    }
}
